presto-main/src/main/java/com/facebook/presto/operator/aggregation/SetOfValues.java

```diff
@@ -0,0 +1,171 @@
+import static java.util.Objects.requireNonNull;
+
+public final class SetOfValues
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(SetOfValues.class).instanceSize();
+    private static final int EXPECTED_ENTRIES = 10;
+    private static final int EXPECTED_ENTRY_SIZE = 16;
+    private static final float FILL_RATIO = 0.75f;
+    private static final int EMPTY_SLOT = -1;
+
+    private final BlockBuilder valueBlockBuilder;
+    private final Type valueType;
+
+    private int[] valuePositionByHash;
+    private int hashCapacity;
+    private int maxFill;
+    private int hashMask;
+
+    public SetOfValues(Type valueType)
+    {
+        this.valueType = requireNonNull(valueType, "valueType is null");
+        valueBlockBuilder = this.valueType.createBlockBuilder(null, EXPECTED_ENTRIES, expectedValueSize(valueType, EXPECTED_ENTRY_SIZE));
+        hashCapacity = arraySize(EXPECTED_ENTRIES, FILL_RATIO);
+        this.maxFill = calculateMaxFill(hashCapacity);
+        this.hashMask = hashCapacity - 1;
+        valuePositionByHash = new int[hashCapacity];
+        Arrays.fill(valuePositionByHash, EMPTY_SLOT);
+    }
+
+    public SetOfValues(Block serialized, Type elementType)
+    {
+        this(elementType);
+        deserialize(requireNonNull(serialized, "serialized is null"));
+    }
+
+    public Block getvalues()
+    {
+        return valueBlockBuilder.build();
+    }
+
+    private void deserialize(Block block)
+    {
+        for (int i = 0; i < block.getPositionCount(); i++) {
+            add(block, i);
+        }
+    }
+
+    public void serialize(BlockBuilder out)
+    {
+        BlockBuilder arrayBlockBuilder = out.beginBlockEntry();
+        for (int i = 0; i < valueBlockBuilder.getPositionCount(); i++) {
+            valueType.appendTo(valueBlockBuilder, i, arrayBlockBuilder);
+        }
+        out.closeEntry();
+    }
+
+    public long estimatedInMemorySize()
+    {
+        long size = INSTANCE_SIZE;
+        size += valueBlockBuilder.getRetainedSizeInBytes();
+        size += sizeOf(valuePositionByHash);
+        return size;
+    }
+
+    /**
+     * Only add this value we haven't seen it before.
+     * Otherwise, ignore it.
+     */
+    public void add(Block value, int valuePosition)
+    {
+        if (!valueExists(value, valuePosition)) {
+            addvalue(value, valuePosition);
+        }
+    }
+
+    private boolean valueExists(Block value, int position)
+    {
+        checkArgument(position >= 0, "position is negative");
+        return valuePositionByHash[getHashPositionOfvalue(value, position)] != EMPTY_SLOT;
+    }
+
+    private void addvalue(Block value, int position)
+    {
+        checkArgument(position >= 0, "position is negative");
+        valueType.appendTo(value, position, valueBlockBuilder);
+        int hashPosition = getHashPositionOfvalue(value, position);
+        if (valuePositionByHash[hashPosition] == EMPTY_SLOT) {
```

Comment:
I know you copied the logic from `KeyValuePairs`. I'm wondering can we simplify this logic? Since this function can only be called when `valuePositionByHash[hashPosition] == EMPTY_SLOT` (this is guarded by logic in `ValueExists`) this should be a `checkState` rather than an `if`. Or if we want to use `if`, we can put the `appendTo` inside the if statement and remove `valueExists`.